client/question/services/code_evaluators/PythonCodeRunnerService.js

```diff
@@ -106,20 +106,74 @@ tie.factory('PythonCodeRunnerService', [
       });
     };
 
+    var _sendCodeToServer = function(code, urlFragment) {
+      clearOutput();
+      return $http({
+        url: SERVER_URL + '/ajax/' + urlFragment,
+        method: 'POST',
+        data: {
+          code: code,
+          language: 'python'
+        }
+      }).then(function(response) {
+        _processServerResponse(response, code, urlFragment);
+      });
+    };
+
+    var _processServerResponse = function(response, code, urlFragment) {
+      var responseData = response.data;
+      if (responseData.stderr.length) {
+        var errorTraceback = ErrorTracebackObjectFactory.fromPythonError(
+          responseData.stderr);
+        return CodeEvalResultObjectFactory.create(
+            code, '', [], [], [], errorTraceback,
+            responseData[VARNAME_MOST_RECENT_INPUT]);
+      } else if (urlFragment === 'run_code') {
```

Comment:
Ah, you shouldn't be passing these urlFragments all over the place...

How about this design instead: compileCodeAsync returns the result of _compileCodeAsync. The latter does `_sendCodeToServer(code, 'compile_code', function() {....})` where the third arg is the successCallback. Ditto for runCodeAsync.

If you like, you can get rid of the _compileCodeAsync and do all this in the publicly-exposed compileCodeAsync itself. Up to you.